xarray/core/array_api_compat.py

```diff
@@ -42,3 +44,29 @@ def result_type(*arrays_and_dtypes, xp) -> np.dtype:
         return xp.result_type(*arrays_and_dtypes)
     else:
         return _future_array_api_result_type(*arrays_and_dtypes, xp=xp)
+
+
+def get_array_namespace(*values):
+    def _get_single_namespace(x):
+        if hasattr(x, "__array_namespace__"):
+            return x.__array_namespace__()
+        elif isinstance(x, array_type("cupy")):
+            # special case cupy for now
+            import cupy as cp
+
+            return cp
+        else:
+            return np
```

Comment:
you probably want to wrap array-api-compat's `array_namespace` here, see https://github.com/scipy/scipy/blob/ec30b43e143ac0cb0e30129d4da0dbaa29e74c34/scipy/_lib/_array_api.py#L118-L152 for what we do in SciPy.

Reply:
Happy to go that route, I did actually try that but array-api-compat doesn't handle a bunch of things we end up passing through this ([scalars](https://github.com/data-apis/array-api-compat/blob/main/array_api_compat/common/_helpers.py#L562), index wrappers, etc) so it would require some careful prefiltering. 

The only things this package effectively wraps that don't have `__array_namespace__` are cupy, dask, and torch. I tried the torch wrapper but it doesn't pass our [as_compatible_data](https://github.com/pydata/xarray/blob/main/xarray/core/variable.py#L265) check because the wrapper object itself doesn't have `__array_namespace__` or `__array_function__` :confused:

Reply:
Looks like [this](https://github.com/data-apis/array-api-compat/blob/main/array_api_compat/common/_helpers.py#L234) would handle the array-like check well, but would require adding this as a core xarray dependency to use it in `as_compatible_data`. Not sure if there is any appetite for that.

Reply:
I played around with this some more.

Getting an object that is compliant from the perspective of `array_api_compat.is_array_api_obj` to work through the hierarchy of xarray objects requires basically swapping in a bunch of more restrictive `hasattr(__array_namespace__)` checks for this function. Not too bad.

I was able to get things to the point that xarray can wrap `torch.Tensor`, which is pretty cool. But the reality is that xarray relies on a lot of functionality beyond the Array API, so this just doesn't work in any practical sense. Similarly, swapping in the more restrictive `cupy.array_api` for the main `cupy` namespace causes all kinds of things to break.

It seems [torch](https://github.com/pytorch/pytorch/issues/58743) has had little movement on supporting the standard since 2021. From xarray's perspective, `cupy` implements everything we need, except for the actual `__array_namespace__` attribute declaring compatibility, which seems to be [planned for v14](https://github.com/cupy/cupy/issues/8269).

So while this compat module is nice in theory, I don't think it's very useful for xarray. cupy, jax, and sparse are good to go without it, and we only need a single special case for cupy to fetch its `__array_namespace__`.